### PR TITLE
Improve efs_facts

### PR DESCRIPTION
**Pull request by willthames**
##### SUMMARY
Avoid an infinite loop when no EFS resources are present

Use standard ansible approaches to pagination, retries,
exception handling, tag processing

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
efs_facts

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 4ac475a967) last updated 2017/10/17 11:49:20 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

### CODE REVIEW

**Comment by willthames** [comment]
Obviously this is a bigger change than required for the avoidance of an infinite loop, but the module meets a lot more of current standards with these fixes.

Ideally there would be tests for efs and efs_facts. I've tested with the EFS resources I have available to me and the tags and targets arguments still work as expected. 

If tests are deemed required then strategy might be:
* write EFS test suite
* create shared module_utils.aws.efs
* update efs_facts and efs to use shared module

**Review by ryansydnor — APPROVED** [review]

File: lib/ansible/modules/cloud/amazon/efs_facts.py [hunk 1]
```diff
@@ -360,17 +343,15 @@ def main():
...
 
     if tags:
-        file_systems_info = filter(lambda item: has_tags(item['Tags'], tags), file_systems_info)
+        file_systems_info = [item for item in file_systems_info if has_tags(item['tags'], tags)]
 
     if targets:
         targets = [(item, prefix_to_attr(item)) for item in targets]
-        file_systems_info = filter(lambda item:
-                                   has_targets(item['MountTargets'], targets), file_systems_info)
+        file_systems_info = [item for item in file_systems_info if has_targets(item['mount_targets'], targets)]
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/efs_facts.py:352]: Can you fix the docs for targets? It's documented as a list of dictionaries, but it needs to just be a list for this to work
```
            targets:
                - "subnet-17bbda4c"
                - "172.31.29.72"
                - [
                    "sg-29d13055",
                    "sg-5e27db2f"
                ]
```
(note that this doesn't work with multiple security groups provided, but that's a preexisting bug and would need has_targets() to be fixed a little).

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/efs_facts.py (hunk 1 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/efs_facts.py:352]: Presumably

```
targets:
  - "subnet-17bbda4c"
  - "172.31.29.72"
  - "sg-29d13055"
  - "sg-5e27db2f"
```

would work though?

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/efs_facts.py:352]: Docs updated as requested.

Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/efs_facts.py:352]: Ah, that does work. :-)

**Review by s-hertel — APPROVED** [review]

**Pull request merged**